File: Binary Search/Search in Rotated Sorted Array/Search in Rotated Sorted Array.kt
class Solution {
    fun search(nums: IntArray, target: Int): Int {
        var left = 0
        var right = nums.size - 1

        while (left <= right) {
            var mid = (left + right) / 2

            var curr = nums[mid]
            if (curr == target) {
                return mid
            } else if (curr < nums[left]) {
                //left-mid is rotated and mid-right is sorted
                if (target > curr && target <= nums[right]) {
                    left = mid + 1
                } else {
                    right = mid - 1
                }
            } else {
                //left-mid is sorted and mid-right is rotated
                if (target < curr && target >= nums[left]) {
                    right = mid - 1
                } else {
                    left = mid + 1
                }
            }
         }
         return -1
    }
}

//Time: O(log n)
//Space Complexity: O(1)

File: Binary Search/Search in Rotated Sorted Array/Search in Rotated Sorted Array.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left, right = 0, len(nums) - 1

        while left < right:
            mid = (left + right) // 2

            curr = nums[mid]

            if curr == target:
                return mid
            elif curr < nums[left]:
                #left-mid is rotated and mid-right is sorted
                if target > curr and target <= nums[right]:
                    left = mid + 1
                else:
                    right = mid - 1
            else:
                if target < curr and target >= nums[left]:
                    right = mid - 1
                else:
                    left = mid + 1
        
        return -1
        
#Time Complexity: O(log n)
#Space Complexity: O(1)

